### ISSUE TITLE
Accent / Combining Mark placement differences between CoreText and own shaper

### ISSUE THREAD

**Issue by drott** [harfbuzz/harfbuzz#1264]
When trying to switch to the own shaping backend on Mac, I see a couple of combining mark placement differences:

1) Lucida Grande Hebrew:

The OT shaper moves the sheva slightly to the left compared to CoreText.

```
$ ./hb-shape /System/Library/Fonts/LucidaGrande.ttc `../test/shaping/hb-unicode-encode U+05E1,U+05B0` --shaper=coretext
[shevahebrew=0+0|samekhhebrew=0+1361]
```

```
$ ./hb-shape /System/Library/Fonts/LucidaGrande.ttc `../test/shaping/hb-unicode-encode U+05E1,U+05B0` --shaper=ot 
[shevahebrew=0@-7,0+0|samekhhebrew=0+1361]
```

2) Lucida Grande m + accent grave

Coretext places U+0300 further north from the m glyph:

```
$ ./hb-shape /System/Library/Fonts/Helvetica.ttc `../test/shaping/hb-unicode-encode U+006D,U+0300` --shaper=coretext
[m=0+1422|gravecmb=0@0,96+283]
```

```
$ ./hb-shape /System/Library/Fonts/Helvetica.ttc `../test/shaping/hb-unicode-encode U+006D,U+0300` --shaper=ot
[m=0+1706|gravecmb=0@-284,10+0]
```

**Comment by brawer** [comment]
Neither `Helvetica.ttc` nor `LucidaGrande.ttc` contain an `ankr` table. Perhaps I’m mis-reading the HarfBuzz code, but does the `kerx` implementation currently fall back to taking anchor points from `glyf` when the font has no `ankr`? The very end of the [kerx spec](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6kerx.html) says:

> If the font has an 'ankr' (anchor point) table, it is used to define the anchor points. Otherwise, the "anchor points" are the points defined by the 'glyf' (glyph data) table.

**Comment by behdad** [comment]
HarfBuzz implements all three ways the anchors can be encoded: https://github.com/harfbuzz/harfbuzz/blob/2e7c716511e8b2cfcd059fa2a2ed4cdd48b351bf/src/hb-aat-layout-kerx-table.hh#L346

I have only tested the one using 'ankr' table though.  I'll test these.  If they use the glyph point method, then this might be that Skia doesn't implement that harfbuzz callback (get_contour_point).

**Comment by behdad** [comment]
HarfBuzz implements all three ways the anchors can be encoded: https://github.com/harfbuzz/harfbuzz/blob/2e7c716511e8b2cfcd059fa2a2ed4cdd48b351bf/src/hb-aat-layout-kerx-table.hh#L346

I have only tested the one using 'ankr' table though.  I'll test these.  If they use the glyph point method, then this might be that Skia doesn't implement that harfbuzz callback (get_contour_point).

**Comment by behdad** [comment]
The Helvetica in your example, does NOT have a GPOS table, and does NOT have a kerx table.  So, no mark positioning in the font.  We do our fallback mark positioning, which moves the mark lower.  Apple doesn't do anything.  So, feature, not bug, in that case.  I'll check the other.

**Comment by behdad** [comment]
Same with LucidaGrande.  Doesn't have kern even.  No positioning info in the font.  We do our fallback.

**Comment by avidrissman** [comment]
In [the Chromium HarfBuzz 2.1 roll CL](https://chromium-review.googlesource.com/c/chromium/src/+/1275945) the [discrepancy with Hebrew vowel positioning](https://chromium-review.googlesource.com/c/chromium/src/+/1275945/34/third_party/WebKit/LayoutTests/platform/mac-mac10.10/fast/text/atsui-spacing-features-expected.png) was linked to this issue, and dismissed as having "been discussed".

In this bug I see discussions about the sheva being positioned slightly to the left or right. The _real_ problem is that the vowels bounce up and down within a word. That's quite wrong. Look at any well-typeset Hebrew text ([random Google result](https://www.tug.org/TUGboat/tb22-3/tb72hoenig-hebrew.pdf)) and you'll see the vowels line up across multiple words at the same height. This important property is lost in that CL roll. Please see [my comment](https://chromium-review.googlesource.com/c/chromium/src/+/1275945#message-019468e1457407780c92ac217410101baa14bea6) on that roll for a link to an annotated snapshot (Google link only though I can provide it).

_Please_ do not ignore this. You _cannot_ treat the vowels at just attached to the individual letters and allow them to be inconsistent across multiple letters. They _have_ to line up across words.

**Comment by behdad** [comment]
This only happen with old and incomplete fonts. Any decent don't controls accent positioning and our fallback logic wouldn't interfere. Cc @jfkthame

**Comment by avidrissman** [comment]
I'm looking at the before-and-after picture from the LayoutTest. LayoutTests/fast/text/atsui-spacing-features.html looks like it uses "Lucida Grande." Did they fix the positioning in the font with 10.11? Whatever CoreText is doing makes it look OK, and I'd definitely call the "after" picture a regression.

**Comment by avidrissman** [comment]
> Any decent don't controls accent positioning

I'm supposing you mean "any decent _font_ controls accent positioning"; please let me know if I misinterpret. How common are fonts that don't, like this font? If this is a one-off problem with Lucida Grande on 10.10, that would be very different than if it's a common problem with many fonts.

**Comment by ebraminio** [comment]
Don't worry Avi, we will try our best :) HarfBuzz is used widely on different OSes and the issue is limited to some fonts and is not like we decide what is better for fonts and font will decide (which considering adoption you can be sure nothing is horribly wrong) but sometimes there are real faults with the fonts (may or may not be with the case) which solving that from our side is somehow hack and we should consider if that hack won't hurt the other correct usages. I've filled https://github.com/harfbuzz/harfbuzz/issues/1356 also which is very similar to this and probably result will help us here also

**Comment by avidrissman** [comment]
I know that HarfBuzz is used all over, but I'm coming from the CL about replacing CT. If CT compensates for broken fonts and we want to replace CT, we may well have to hack around broken fonts too.

**Comment by ebraminio** [comment]
Hmm, disabling hb-ot-shape-fallback#position_mark makes Lucida Grandeto work as expected but losing Helvetica's grave, I supposed to think theu use similar logic in lack of kern, but they don't?

**Comment by ebraminio** [comment]
```
➜  harfbuzz git:(lcar) ✗ util/hb-shape LucidaGrande.ttc -u U+05E1,U+05B0 --shaper ot --font-funcs=ft 
[shevahebrew=0@4,0+0|samekhhebrew=0+1361]
➜  harfbuzz git:(lcar) ✗ util/hb-shape LucidaGrande.ttc -u U+05E1,U+05B0 --shaper ot --font-funcs=ot
[shevahebrew=0@-7,0+0|samekhhebrew=0+1361]
```

Interesting... why it is different between ft and ot font-funcs?

But https://chromium-review.googlesource.com/c/chromium/src/+/1275945/34/third_party/WebKit/LayoutTests/platform/mac-mac10.10/fast/text/atsui-spacing-features-expected.png is way out, maybe something to be improved with Chrome font funcs?

**Comment by behdad** [comment]
> Interesting... why it is different between ft and ot font-funcs?

That's hb-ft issue.  Ignore.

**Comment by ebraminio** [comment]
https://chromium-review.googlesource.com/c/chromium/src/+/1275945/34/third_party/WebKit/LayoutTests/platform/mac-mac10.10/fast/text/atsui-spacing-features-expected.png difference (at below) feels way higher than this numbers.

![image](https://user-images.githubusercontent.com/833473/48073639-6c4c2700-e1f4-11e8-9e11-fb109fbae1d6.png)

The one on right is HarfBuzz and the difference is unnoticeable.
 
```
util/hb-view /System/Library/Fonts/LucidaGrande.ttc -u U+05d9,U+05b0,U+05d4,U+05b4,U+05d9,U+05d0,U+05b8,U+05d7,U+05b4,U+05d9,U+05dc,U+05b0,U+05da,U+05b8,U+05e1,U+05b5,U+05e4,U+05b6,U+05e8,U+05e9,U+05c1,U+05b0,U+05dc,U+05b7,U+05d7,U+05b0,U+05ea,U+05bc,U+05b4,U+05d9,U+05d5,U+05d5,U+05bc,U+05de,U+05b4,U+05de,U+05b0,U+05db,U+05bc,U+05b6,U+05e8,U+05b6,U+05ea,U+05e6,U+05b0,U+05de,U+05b4,U+05d9,U+05ea,U+05d5,U+05bc,U+05ea,U+05dc,U+05b8,U+05da,U+05b0,U+05de,U+05b0,U+05db,U+05b7,U+05e8,U+05b0,U+05ea,U+05bc,U+05b4,U+05d9,U+05d5 --shaper ot --font-size 20

util/hb-view /System/Library/Fonts/LucidaGrande.ttc -u U+05d9,U+05b0,U+05d4,U+05b4,U+05d9,U+05d0,U+05b8,U+05d7,U+05b4,U+05d9,U+05dc,U+05b0,U+05da,U+05b8,U+05e1,U+05b5,U+05e4,U+05b6,U+05e8,U+05e9,U+05c1,U+05b0,U+05dc,U+05b7,U+05d7,U+05b0,U+05ea,U+05bc,U+05b4,U+05d9,U+05d5,U+05d5,U+05bc,U+05de,U+05b4,U+05de,U+05b0,U+05db,U+05bc,U+05b6,U+05e8,U+05b6,U+05ea,U+05e6,U+05b0,U+05de,U+05b4,U+05d9,U+05ea,U+05d5,U+05bc,U+05ea,U+05dc,U+05b8,U+05da,U+05b0,U+05de,U+05b0,U+05db,U+05b7,U+05e8,U+05b0,U+05ea,U+05bc,U+05b4,U+05d9,U+05d5 --shaper coretext --font-size 20
```

Just that T like shape vowel (left most occurrence) is different which HB looks better even. Something else is going on.

**Comment by ebraminio** [comment]
No different on its 10.10 version also,

CoreText is on right,

![image](https://user-images.githubusercontent.com/833473/48074972-5d1aa880-e1f7-11e8-95c2-39f6a577a21d.png)

`Lucida Grande; 9.0d11e1; 2014-04-02`

So I believe this is from Chrome side.

**Comment by ebraminio** [comment]
10.13 Lucida Grande, no difference, harfbuzz master left, CoreText right.

![image](https://user-images.githubusercontent.com/833473/48182252-c3aade00-e33f-11e8-8329-c4452794f4a5.png)

**Comment by avidrissman** [comment]
+1! Thank you so much for getting this fixed.

**Comment by ebraminio** [comment]
To be honest we haven't figured the issue cause yet just we know is not from HarfBuzz :/

**Comment by drott** [comment]
behdad@, contrary to what we discussed on Friday, and contrary to what I believe you meant in the comment above: https://github.com/harfbuzz/harfbuzz/issues/1264#issuecomment-430863120

For both,

```
./hb-shape /System/Library/Fonts/LucidaGrande.ttc --shaper=ot --font-funcs=ot -u 006D,0300
```
as well as

```
./hb-shape /System/Library/Fonts/Helvetica.ttc --shaper=ot --font-funcs=ot -u 006D,0300
```

the plan flag `fallback_mark_positioning` is false, and line 881 of `hb-ot-shape.cc`, where I believe fallback mark positioning is happening, is skipped:

```
  if (c->plan->fallback_mark_positioning && c->plan->shaper->fallback_position)
    _hb_ot_shape_fallback_mark_position (c->plan, c->font, c->buffer);
```

If I make a subsetted font `Helv1.subset.ttf` from Helvetica, which is subsetted for `m` and `gravecmb`, which drops `mort` and `kern`, and then run this in the debugger, the plan flag fallback_mark_positioning is true, and the combining mark placement above the m looks much better. Also, I do see callbacks to `hb_ot_get_glyph_extents`, whereas when the plan flag is false, I do not see those.

I started to investigate this when I did not see any glyph-extents callbacks being invoked in Blink for those test cases where we have issues with combining mark placements (the Helvetica m + gravecmb, and the LucidaGrande Hebrew tests).

```
$ ./hb-view Helv1.subset.ttf  --shaper=ot --font-funcs=ot  -u 006D,0300 > m_subset.png
```

![m_subset](https://user-images.githubusercontent.com/1021784/48777048-cdcec400-ecd1-11e8-826f-613acbb615cf.png)


```
$ ./hb-view Helvetica.ttc  --shaper=ot --font-funcs=ot  -u 006D,0300 > m_full_helvetica.png
```

![m_full_helvetica](https://user-images.githubusercontent.com/1021784/48777056-d1fae180-ecd1-11e8-98f3-8e9976ab462e.png)
